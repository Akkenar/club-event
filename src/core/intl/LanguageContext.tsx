import * as React from 'react';
import { IntlContext } from './intl.type';

const LanguageContext = React.createContext<IntlContext>({
  confirmation: '',
  handleChangeLanguage: () => {
    /* noop */
  },
  information: '',
  messages: {},
});

export default LanguageContext;
