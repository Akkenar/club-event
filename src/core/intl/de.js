import messages from '../../../translations/de/messages.de.json';
import confirmation from '../../../translations/de/confirmation.de.md';
import information from '../../../translations/de/information.de.md';

const intl = {
  messages,
  confirmation,
  information,
};

export default intl;
