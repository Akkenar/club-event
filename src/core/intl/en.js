import messages from '../../../translations/en/messages.json';
import confirmation from '../../../translations/en/confirmation.md';
import information from '../../../translations/en/information.md';

const intl = {
  messages,
  confirmation,
  information,
};

export default intl;
