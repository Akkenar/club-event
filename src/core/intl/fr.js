import messages from '../../../translations/fr/messages.json';
import confirmation from '../../../translations/fr/confirmation.md';
import information from '../../../translations/fr/information.md';

const intl = {
  messages,
  confirmation,
  information,
};

export default intl;
