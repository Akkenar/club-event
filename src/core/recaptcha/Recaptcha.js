import React, { useEffect, useState } from 'react';

const CONTAINER_ID = 'recaptcha-container';
const SCRIPT_ID = 'recaptcha-script';

function appendRecaptchaScript() {
  const script = document.createElement('script');
  script.type = 'text/javascript';
  script.id = SCRIPT_ID;
  script.src =
    'https://www.google.com/recaptcha/api.js?onload=onloadCallback&render=explicit';
  document.head.appendChild(script);
}

function renderCaptcha() {
  // Render the captcha
  grecaptcha.render(CONTAINER_ID, {
    sitekey: '6LeZ23kUAAAAALdlAuJg3X0MTmPelUzvJ4dAMpK-',
  });
}

const Recaptcha = () => {
  const [loaded, setLoaded] = useState(false);
  useEffect(() => {
    // In case the script is not loaded.
    if (!document.getElementById(SCRIPT_ID)) {
      // Global callback hooked to the recaptcha script src to only render
      // the captcha when the script is loaded.
      window.onloadCallback = () => {
        renderCaptcha();
        setLoaded(true);
      };

      // Append the script.
      appendRecaptchaScript();
    } else if (!loaded) {
      // Direct render.
      renderCaptcha();
      setLoaded(true);
    }
  });

  return <div id={CONTAINER_ID} className="g-recaptcha" />;
};

export default Recaptcha;
