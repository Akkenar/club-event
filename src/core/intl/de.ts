import * as confirmation from '../../../translations/de/confirmation.md';
import * as information from '../../../translations/de/information.md';
import * as messages from '../../../translations/de/messages.json';
import { IntlType } from './intl.type';

const intl: IntlType = {
  confirmation,
  information,
  messages: (messages as any).default,
};

export default intl;
