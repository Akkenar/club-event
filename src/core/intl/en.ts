import confirmation from '../../../translations/en/confirmation.md';
import information from '../../../translations/en/information.md';
import { default as messages } from '../../../translations/en/messages.json';
import { IntlType } from './intl.type';

const intl: IntlType = {
  confirmation,
  information,
  messages,
};

export default intl;
