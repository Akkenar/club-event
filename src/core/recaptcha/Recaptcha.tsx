import React from 'react';
import { useContext } from 'react';
import getKey from '../intl/getKey';
import LanguageContext from '../intl/LanguageContext';
import { useIntersectionObserver } from '../useIntersectionObserver';
import { useCaptchaScript } from './useCaptchaScript';

const CAPTCHA_SIZE = { height: 78, width: 302 };

const Recaptcha = () => {
  const { isDisplayed, startObserving } = useIntersectionObserver();
  const { messages, language } = useContext(LanguageContext);
  const { isLoaded, isScriptAdded, containerId } = useCaptchaScript(
    language,
    isDisplayed,
  );

  // If no language, don't bother rendering the captcha, just the placeholder
  if (!language) {
    return <div style={CAPTCHA_SIZE}>&nbsp;</div>;
  }

  // Lazy loading, if enabled.
  if (!isDisplayed) {
    return (
      <div
        style={CAPTCHA_SIZE}
        ref={startObserving}
        data-testid="captcha-loading"
      >
        {getKey('loading', messages)}
      </div>
    );
  }

  if (!isScriptAdded) {
    // Nothing, so when we change the language, the element in the DOM is deleted.
    return null;
  }

  // Container for the captcha, the ID matters.
  return (
    <div
      style={CAPTCHA_SIZE}
      id={containerId}
      data-testid={`recaptcha--${isLoaded}`}
    >
      {getKey('loading', messages)}
    </div>
  );
};

export default Recaptcha;
