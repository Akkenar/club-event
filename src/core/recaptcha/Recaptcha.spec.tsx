import * as React from 'react';
import { Fragment } from 'react';
import { render, waitForElement } from 'react-testing-library';
import { mockGlobalProperty } from '../../test-utils/test-utils.lib';
import { IntlContext } from '../intl/intl.type';
import LanguageContext from '../intl/LanguageContext';
import Recaptcha from './Recaptcha';

const languageContextValue: IntlContext = {
  confirmation: '',
  information: '',
  language: 'en',
  messages: {},
};

const RecaptchaWithLanguage = () => (
  <LanguageContext.Provider value={languageContextValue}>
    <Recaptcha />
  </LanguageContext.Provider>
);

// Simulate the script loading.
jest.spyOn(document.head, 'appendChild').mockImplementation(() => {
  (window as any).onloadCallback();
});

mockGlobalProperty(window)('grecaptcha')({
  render: jest.fn(),
});

describe('Recaptcha', () => {
  it('should render the recaptcha', async () => {
    const wrapper = render(<RecaptchaWithLanguage />);
    await waitForElement(() => wrapper.getByTestId('recaptcha--true'));
    expect((window as any).grecaptcha.render).toHaveBeenCalledTimes(1);
    expect(wrapper.baseElement).toMatchSnapshot();
  });
});
