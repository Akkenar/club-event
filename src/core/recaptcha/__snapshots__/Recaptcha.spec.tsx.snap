// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recaptcha should render the recaptcha 1`] = `
<body>
  <div>
    <div
      class="g-recaptcha"
      data-testid="recaptcha--true"
      id="recaptcha-container"
      style="height: 78px; width: 302px;"
    >
      loading
    </div>
  </div>
</body>
`;
